src/client.ts

```diff
@@ -1,3 +1,106 @@
+import { PlayerInfo, SessionInfo } from "./types";
+import axios from 'axios'
+
 export default class ArcadeClientSDK {
-  
+
+  sessionToken: string = ""
+  url = "https://userapi.ultimatearcade.io"
+  gotEndResponse = false
+
+  constructor(options?: {
+    baseDomain?: string
+  }) {
+    // Register listener for iframe messaging
+    window.onmessage = (e) => {
+      // Only set this once
+      if (this.sessionToken === "" && e.data.token) {
+        this.sessionToken = e.data.token
+      }
+      if (e.data.token) {
+        // Send back confirmation regardless
+        window.top?.postMessage({msg: 'gotToken'}, '*')
+      }
+      if (e.data.msg && e.data.msg === "gameOver") {
+        // The parent window confirming game over received
+        this.gotEndResponse = true
+      }
+    }
+
+    if (options?.baseDomain) {
+      this.url = `https://userapi.${options.baseDomain}`
+    }
+  }
+
+  /**
+   * Fetches the information about the game session, most importantly the address of the game server to connect to
+   */
+  getSessionInfo(): SessionInfo {
+    if (this.sessionToken === "") {
+      throw new Error("parent page not ready yet, try again in a bit")
+    }
+
+    // Decode the JWT
+    const decoded = JSON.parse(Buffer.from(this.sessionToken.split(".")[1], "base64").toString())
+
+    return {
+      server_address: decoded.addr
+    }
+  }
+
+  /**
+   * Fetches profile information about the player, like their display name
+   */
+  async getPlayerProfile(): Promise<PlayerInfo> {
+    try {
+      let statuscode = 0
+      let retryCount = 0
+      let data: PlayerInfo | null = null
+      while (statuscode !== 200) {
+        if (retryCount > 0) {
+          await new Promise((r) => {
+            setTimeout(() => {
+              r(null)
+            }, 200 * retryCount)
+          })
+        }
+        const res = await axios.post(this.url+'/games/player-profile', {}, {
+          headers: {
+            'Authorization': `Bearer ${this.sessionToken}`
+          }
+        })
+        statuscode = res.status
+        if (statuscode > 299 && statuscode < 500) {
+          throw new Error(`server returned ${statuscode} - ${res.statusText}`)
+        }
+
+        if (statuscode <= 299) {
+          data = res.data
+        }
+
+        retryCount++
+        if (retryCount >= 5) {
+          throw new Error(`max retries hit, exiting with status code ${statuscode}`)
+        }
+      }
+      return data as PlayerInfo
+    } catch (error) {
+      throw error
+    }
+  }
+
+  /**
+   * Tells the Ultimate Arcade parent window that the game session has ended, and the game client can be removed from the page and the results shown.
+   */
+  async gameOver(): Promise<void> {
+    while (!this.gotEndResponse) {
+      window.top?.postMessage({msg: 'gameOver'}, '*')
+
+      // Sleep 50ms before checking again
+      await new Promise((r) => {
+        setTimeout(() => {
+          r(null)
+        }, 50)
+      })
```

Comment:
why so complicated? The parent window is always there and can register its handlers before the iframe loads, so no reason to complicate things with retries here

Reply:
Alright, but I generally go for retries everywhere... better to have and not need than the inverse

Reply:
I'd agree if this was a network call, but it isn't. This might lead to quite some confusion down the line: E.g. we could receive old messages from another iframe and not properly distinguish between them.
Actually thinking about this, it's a bit hard to figure out just by looking at the origin property to what iframe these messages relate to. We might need to introduce some random ID for the onMessage handler to properly differentiate them

Reply:
we could pass the token back in every message since that is unique to a game session?

Reply:
Or do you mean PER message

Reply:
alright then we can just use the session token, will include in the messages

Reply:
@Hades32 added